drivers/input/input_tsc_keys.c

```diff
@@ -0,0 +1,174 @@
+		LOG_ERR("TSC@%p: No child config for group %d", config->tsc_dev, group);
+		return;
+	}
+
+	struct input_tsc_keys_child_data *child_data = child_config->data;
+
+	if (ring_buf_item_space_get(&child_data->rb) == 0) {
+		uint32_t oldest_point;
+		int32_t slope;
+
+		(void)ring_buf_get(&child_data->rb, (uint8_t *)&oldest_point, sizeof(oldest_point));
+
+		slope = value - oldest_point;
+		if (slope < -child_config->noise_threshold &&
+		    child_data->state == INPUT_TSC_KEYS_EXPECT_DOWN) {
+			/* this is a key press, now expect a release */
+			child_data->state = INPUT_TSC_KEYS_EXPECT_UP;
+			input_report_key(dev, child_config->zephyr_code, 1, false, K_NO_WAIT);
+		} else if (slope > child_config->noise_threshold &&
+			   child_data->state == INPUT_TSC_KEYS_EXPECT_UP) {
+			/* this is a key release, now expect a press */
+			child_data->state = INPUT_TSC_KEYS_EXPECT_DOWN;
+			input_report_key(dev, child_config->zephyr_code, 0, false, K_NO_WAIT);
+		}
+	}
+
+	(void)ring_buf_put(&child_data->rb, (uint8_t *)&value, sizeof(value));
+}
+
+int input_tsc_keys_init(const struct device *dev)
+{
+	const struct input_tsc_keys_config *config = dev->config;
+	struct input_tsc_keys_data *data = dev->data;
+
+	if (!device_is_ready(config->tsc_dev)) {
+		LOG_ERR("TSC@%p: TSC device not ready", config->tsc_dev);
+		return -ENODEV;
+	}
+
+	for (uint8_t i = 0; i < config->child_num; i++) {
+		const struct input_tsc_keys_child_config *child = &config->children[i];
+		struct input_tsc_keys_child_data *child_data = child->data;
+
+		ring_buf_item_init(&child_data->rb, CONFIG_INPUT_TSC_KEYS_BUFFER_WORD_SIZE,
+				   child_data->buffer);
+	}
+
+	stm32_tsc_register_callback(config->tsc_dev, config->tsc_callback);
+
+	k_timer_init(&data->sampling_timer, input_tsc_sampling_timer_callback, NULL);
+	k_timer_user_data_set(&data->sampling_timer, (void *)config->tsc_dev);
+	k_timer_start(&data->sampling_timer, K_NO_WAIT, K_MSEC(config->sampling_interval_ms));
+
+	return 0;
+}
+
+#define TSC_KEYS_CHILD(node_id)                                                                    \
+	{                                                                                          \
+		.label = DT_PROP(node_id, label),                                                  \
+		.group_index = DT_PROP(DT_PHANDLE(node_id, group), group),                         \
+		.zephyr_code = DT_PROP(node_id, zephyr_code),                                      \
+		.noise_threshold = DT_PROP(node_id, noise_threshold),                              \
+		.data = &tsc_keys_child_data_##node_id,                                            \
+	},
+
+#define TSC_KEYS_CHILD_DATA(node_id)                                                               \
+	static struct input_tsc_keys_child_data tsc_keys_child_data_##node_id;
+
+#define TSC_KEYS_INIT(node_id)                                                                     \
+	DT_INST_FOREACH_CHILD_STATUS_OKAY(node_id, TSC_KEYS_CHILD_DATA)                            \
```

Comment:
would it work to create an array of `input_tsc_keys_child_data` sized by `DT_CHILD_NUM`? save few pointers

Reply:
made it much simpler, thanks